### ISSUE TITLE
Data Generator from file with batches

### ISSUE THREAD

**Issue by kramorov** [kramorov/kramorov#2]
class DataGenerator(tf.compat.v2.keras.utils.Sequence):
  def __init__(self, path_to_images, path_to_segmentation, image_files_list, segmentation_files_list, number_of_classes, batch_size, img_width, img_height):        
    
    self.path_to_images = path_to_images
    self.path_to_segmentation = path_to_segmentation
    self.image_files_list=image_files_list
    self.segmentation_files_list=segmentation_files_list

    self.img_width = img_width
    self.img_height = img_height

    self.batch_size = batch_size
    self.number_of_classes = number_of_classes

    self.cur_batch_num = 0
    self.num_of_batches=math.ceil(len(self.image_files_list) / self.batch_size)

  def __next__(self):
    data = self.__getitem__(self.cur_batch_num)  # Получим очередной пакет
    self.cur_batch_num += 1 # Номер следующего пакета
    if self.cur_batch_num >= self.__len__(): # Если мы обработали все пакеты в датасете, то:
        # self.on_epoch_end
        self.cur_batch_num = 0
    return data    
  def __len__(self):
    # Возвращаем количество пакетов в датасете
    return self.num_of_batches    
  def __getitem__(self, batch_index):
    x_temp = self.get_x_train_batch(batch_index=batch_index)
    y_temp = self.get_y_train_batch(batch_index=batch_index)
    #print(f'Files = {self.image_files_list[file_index]}, {self.segmentation_files_list[file_index]}; x_shape={x_temp.shape}, y_shape={y_temp.shape}')
    return x_temp, y_temp

  def on_epoch_end(self):
    # В конце эпохи перемешаем названия файлов
    #seed = np.random.randint()
    #self.x = shuffle(self.image_files_list, random_state=seed)
    #self.y = shuffle(self.segmentation_files_list, random_state=seed)
    random.shuffle(self.image_files_list)
    random.shuffle(self.segmentation_files_list)
    return

  def get_x_train_batch(self, batch_index):
    batch_x = self.image_files_list[batch_index * self.batch_size : (batch_index + 1) * self.batch_size]
    x_data = list()
    for file_name in batch_x:
      try:
        full_path_to_file=os.path.join(self.path_to_images + '/',file_name)
        current_image=image.load_img(full_path_to_file, target_size=(self.img_width, self.img_height))
        x = image.img_to_array(current_image)
        x_data.append(x)
        #print(f'Загружен файл {full_path_to_file}')
      except Exception as e:
        print(e)
        print(f'Ошибка загрузки файла {full_path_to_file}')
    x_array_norm=np.array(x_data, dtype='uint8')
    return x_array_norm
  
  def get_y_train_batch(self, batch_index):
    batch_y = self.segmentation_files_list[batch_index * self.batch_size : (batch_index + 1) *self.batch_size]
    y_data = list()
    for file_name in batch_y:
      try:
        full_path_to_file = os.path.join(self.path_to_segmentation + '/',file_name)
        current_sermentation=image.load_img(full_path_to_file, target_size=(self.img_width, self.img_height))
        y = image.img_to_array(current_sermentation)
        y = self.rgbToohe(y) # Получаем OHE-представление сформированного массива
        y_data.append(y)
        #print(f'Загружен файл {full_path_to_file}')
      except Exception as e:
        print(e)
        print(f'Ошибка загрузки файла {full_path_to_file}')
    return np.array(y_data) 

  # Функция преобразования пикселя сегментированного изображения в индекс (6 классов)
  def color2index(self, color):
    index=0
    if (color[0] + color[1] + color[2]) > 20  : index = 1 # самолет    
    return index   
  # Функция перевода индекса пикселя в to_categorical
  def rgbToohe(self, y): 
    y_shape = y.shape # Запоминаем форму массива для решейпа
    y = y.reshape(y.shape[0] * y.shape[1], 3) # Решейпим в двумерный массив
    yt = [] # Создаем пустой лист
    for i in range(len(y)): # Проходим по всем трем канала изображения
        yt.append(utils.to_categorical(self.color2index(y[i]), num_classes= self.number_of_classes)) # Переводим пиксели в индексы и преобразуем в OHE
    yt = np.array(yt) 
    yt = yt.reshape(y_shape[0], y_shape[1], self.number_of_classes) # Решейпим к исходныму размеру
    return yt
